File: tests/cpp/test_complexMixer.cpp
#include <cmath>
#include <gtest/gtest.h>
#include "utils.h"
#include "dsp.h"
#include "complexMixer.h"

using namespace std;

class TestComplexMixer : public ::testing::Test {
    protected:
        dsp::ComplexFFT fft;
        ComplexMixer complexMixer;
};

TEST_F(TestComplexMixer, validOutput) {
    double fshift = 250.0;
    double fs = 1000.0;
    double f = 100.0;
    vector<double> I = utils::linspace(0.0, 1.0, static_cast<uint>(fs), false);
    for (uint i = 0; i < I.size(); i++) { I[i] = cos(2 * M_PI * f * I[i]); }
    vector<double> Q(static_cast<uint>(fs), 0);

    vector<double> Iout = complexMixer(fshift, fs, I, Q);
    vector<complex<double>> fftIout = fft(Iout);
    ASSERT_EQ(fftIout.size(), I.size()) << "Invalid size of the shifted signal.";
    for (uint i = 0; i < fftIout.size() / 2; i++) {
        double sampleExpected = 0.0;
        if (i == static_cast<int>(fshift-f) || i == static_cast<int>(fshift+f)) { sampleExpected = fs / 4; }
        ASSERT_NEAR(abs(fftIout[i]), sampleExpected, 1e-5) << "Spectral components must be at fshift-f and fshifted+f and noise close to 0.";
    }
}

TEST_F(TestComplexMixer, invalidInput) {
    double fs = -1000.0; // Sampling frequency must be positive.
    EXPECT_THROW(complexMixer(100.0, fs, {1}, {1}), invalid_argument);
}

File: tests/python/test_complex_mixer.py
import numpy as np
import pytest
from complex_mixer import ComplexMixer

@pytest.fixture
def complex_mixer():
    return ComplexMixer()

def test_complex_mixer_valid_output(complex_mixer):
    f_shift = 250
    fs = 1000
    f = 100
    t = np.linspace(0, 1, fs, endpoint=False)
    I = np.cos(2 * np.pi * f * t)
    Q = np.zeros_like(t)

    I_out = complex_mixer(f_shift, fs, I, Q)
    assert(I_out.size == I.size), "Invalid size of the shifted signal."
    I_out = np.abs(np.fft.fft(I_out))[:len(I_out)//2]
    assert(np.allclose(I_out[[f_shift-f, f_shift+f]], fs // 4, atol=1e-5)), "Spectral components must be at fshift-f and fshifted+f."
    assert(np.allclose(I_out[(I_out != I_out[f_shift-f]) & (I_out != I_out[f_shift+f])], 0, atol=1e-5)), "Noise must be close to 0."

def test_complex_mixer_invalid_input(complex_mixer):
    with pytest.raises(ValueError, match="ComplexMixer.__call__: Sampling frequency must be positive."):
        complex_mixer(200.0, -1000, np.array([1]), np.array([1]))
